[326] Configurable preview route

juanmanuelramallo opened the pull request:
Resolves #326 
<!-- See https://github.com/github/view_component/blob/master/CONTRIBUTING.md#submitting-a-pull-request  -->

### Summary

- Added a config attribute to allow users to configure the endpoint to be used for previewing components
- Updated views to use a helper path method instead of hardcoded paths
- Updated the engine to use the new configuration attribute
- Removed unused "trim_result" method from test helper
- Added a test helper method to run assertions with a different preview route config
- Updated integration test

<!-- Provide a general description of the code changes in your pull
request... were there any bugs you had fixed? If so, mention them. If
these bugs have open GitHub issues, be sure to tag them here as well,
to keep the conversation linked together. -->

### Other Information

I'm not entirely convinced about naming this config `preview_route`. I hear suggestions. 
<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file. -->

jonspalmer commented on app/controllers/view_components_controller.rb:
```diff
@@ -59,4 +61,8 @@ def set_locale
+
+  def url_helpers
+    Rails.application.routes.url_helpers
+  end
```
Don't all controllers have url_helpers already? I'm surprised we need this.

juanmanuelramallo replied:
Surprisingly enough, Rails::ApplicationController (parent class) does not have the url helpers. 

Maybe this is a sign that we should start to decouple from rails internals, like the Rails::ApplicationController and start rolling our own standalone controller (?)

jonspalmer commented on app/views/view_components/index.html.erb:
```diff
@@ -1,8 +1,8 @@
   <ul>
     <% preview.examples.each do |preview_example| %>
-      <li><%= link_to preview_example, "/rails/view_components/#{preview.preview_name}/#{preview_example}" %></li>
+      <li><%= link_to preview_example, preview_view_component_path("#{preview.preview_name}/#{preview_example}") %></li>
```
nice

jonspalmer commented on README.md:
````diff
@@ -474,6 +474,15 @@ Preview classes live in `test/components/previews`, which can be configured usin
+
+`config/application.rb`
+```ruby
+config.view_component.preview_route = "/previews"
````
I think we'd call this `preview_path` ?

juanmanuelramallo replied:
We already have that config for enabling users configure where the previews live in the codebase. Should we rename?

jonspalmer replied:
oh right. hmm... I'd say `preview_source_path` and `preview_path` but that's a confusing breaking change. Let's debate the mount idea before we are forced to decide on this.

joelhawksley replied:
I prefer this approach to the mounted engine implementation. I also think `preview_route` is fine 👍

joelhawksley commented on app/controllers/view_components_controller.rb:
```diff
@@ -59,4 +61,8 @@ def set_locale
     prepend_view_path Rails.root.join("app/views") if defined?(Rails.root)
   end
+
+  def url_helpers
+    Rails.application.routes.url_helpers
+  end
```
Do we need these helpers for anything else? If we don't, what if we just defined `preview_component_path` directly?

```suggestion
  def preview_component_path
    Rails.application.routes.url_helpers.preview_component_path
  end
```

juanmanuelramallo replied:
Great call! Will update

jonspalmer replied:
I'm still wondering if we do controllers as part of the engine can we get the URL helpers for free event if we go with this configuration mechanism vs mounting the engine?

juanmanuelramallo replied:
As long as our controller inherits from Rails::ApplicationController the URL helpers will be needed

juanmanuelramallo replied:
Hmmm just did some testing and turns out we don't need it, I wonder why I added it in the first place.

joelhawksley commented on README.md:
````diff
@@ -474,6 +474,15 @@ Preview classes live in `test/components/previews`, which can be configured usin
 config.view_component.preview_path = "#{Rails.root}/lib/component_previews"
 ```
 
+Previews are served from <http://localhost:3000/rails/view_components> by default. If you want to use a different endpoint you can configure the `preview_route` option:
````
```suggestion
Previews are served from <http://localhost:3000/rails/view_components> by default. To use a different endpoint, set the `preview_route` option:
```

joelhawksley commented on lib/view_component/engine.rb:
```diff
@@ -14,6 +14,7 @@ class Engine < Rails::Engine # :nodoc:
 
       if options.show_previews
         options.preview_path ||= defined?(Rails.root) ? "#{Rails.root}/test/components/previews" : nil
+        options.preview_route ||= "/rails/view_components"
```
```suggestion
```

joelhawksley commented on lib/view_component/previewable.rb:
```diff
@@ -20,6 +20,14 @@ module Previewable
+      #
+      # Defaults to +/rails/view_components+ when `show_previews' is enabled
+      #
+      mattr_accessor :preview_route, instance_writer: false
```
```suggestion
      mattr_accessor :preview_route, instance_writer: false, default: "/rails/view_components"
```

(see https://apidock.com/rails/Module/mattr_accessor)

juanmanuelramallo replied:
Nice

joelhawksley commented on lib/view_component/engine.rb:
```diff
@@ -64,8 +65,8 @@ class Engine < Rails::Engine # :nodoc:
         app.routes.prepend do
-          get "/rails/view_components"       => "view_components#index", :internal => true
-          get "/rails/view_components/*path" => "view_components#previews", :internal => true
+          get options.preview_route, to: "view_components#index", as: "preview_view_components", internal: true
+          get "#{options.preview_route}/*path", to: "view_components#previews", as: "preview_view_component", internal: true
```
```suggestion
          get options.preview_route, to: "view_components#index", as: :preview_view_components, internal: true
          get "#{options.preview_route}/*path", to: "view_components#previews", as: :preview_view_component, internal: true
```

juanmanuelramallo commented:
Rails 5.0.0 had a different implementation for the default in mattr_accessor it seems https://api.rubyonrails.org/v5.0.0/classes/Module.html#method-i-mattr_accessor 
It receives a block instead of a kwarg

juanmanuelramallo commented:
@jonspalmer @joelhawksley would you mind giving it another look? I added Joel's suggestions and added two commits, one to fix the engine configuration and the other to make it work on Rails <6.

joelhawksley commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 # master
 
+* Add `config.view_component.preview_route` to set the endpoint for component previews. By default `/rails/view_component` is used.
```
```suggestion
* Add `config.view_component.preview_route` to set the endpoint for component previews. By default `/rails/view_components` is used.
```
